README.md
# custom_dropdown

A Flutter Package that provides you an beautified alternative of Dropdown widget.
It is highly Customisable in terms of size, color, style, etc.


## Import

``` yaml
dependencies:
  custom_picker: <latest_version>
```


## Output
![custom_picker](https://github.com/dubey2k/custom_picker/blob/master/example/output/output.gif)


## Usage

You can use multiple text beautification options with detected text formats.

```dart
CustomPicker(
  // deciding list whether to show loop over popup items
  optionLoop: [false, false],
  // give popup height
  popupHeight: 200,
  // where to show popupmenu
  below: false,
  // give where to place popup vertically
  verticalOffset: 10,
  // hanfle all changes here
  handleChange: (list) {
    print(list);
  },
  // to include yes no dialog or not
  yesNo: true,
  //give color to popup selected option
  popUpSelColor: Colors.black.withAlpha(150),
  // customize text style for each popUp list
  popUpTextStyle: const [
    TextStyle(color: Colors.white, fontSize: 20),
    TextStyle(color: Colors.white, fontSize: 20),
  ],
  // customize text style for each list
  widTextStyle: const [
    TextStyle(color: Colors.black, fontSize: 18),
    TextStyle(color: Colors.black, fontSize: 18),
  ],
  // customize divider color
  divColor: Colors.black,
  //give a decoration to your popupmenu
  popUpDecoration: BoxDecoration(
      color: Colors.black.withAlpha(100),
      borderRadius: BorderRadius.circular(10)),
  // decorate your picker widget
  widDecoration: BoxDecoration(
    border: Border.all(width: 2, color: Colors.black),
    borderRadius: BorderRadius.circular(10),
  ),
  // input all list you want to display in popupmenu
  list: const [
    ['1', '2', '3', '4', '5', '6'],
    ['one', 'two', 'three']
  ],
  // control flex property of picker children
  wtList: [3, 1],
  // callback that'll be call when option of YesNo dialog is clicked
  onChanged: (List<int> value) {
    print(value);
  },
  //give initial selected item indicies
  selected: const [0, 1],
)
       
```


## Parameters


|   parameter        |      description                                                                                     |  default                                 |
| ------------------ | ---------------------------------------------------------------------------------------------------- | ---------------------------------------- |
|   popupHeight      |     The height of the popup                                                                          |  400                                     |
|   optionLoop       |     List of boolswhich child to loop over wheel                                                      |  true for all children                   |
|   below            |     show popup widget below the widget                                                               |  Color(0xFF424242)                       |
|   popUpDecoration  |     The FontWeight of UnDetected Text                                                                |  radius : 10, border: Color(0xFF424242)  |
|   verticalOffset   |     vertical position of popup widget                                                                |  TextAlign.start                         |
|   selected         |     initially selected list                                                                          |  required                                |
|   yesNo            |     whether to show YesNo dialog                                                                     |  false                                   |
|   popUpTextStyle   |     add TextStyle of popup children                                                                  |  TextStyle(fontSize: 17)                 |
|   widTextStyle     |     add TextStyles of Picker Children                                                                |  TextStyle()                             |
|   onChanged        |     Whenever the current date is changed. If this is null, the picker is considered disabled         |  required                                |
|   handleChange     |     any change in any child will call this method                                                    |  void                                    |
|   popUpMargin      |     Margin for PopUp Widget (if it is zero, pop up widget will be visible at the center of picker)   |  hor = 20 , ver =20                      |
|   onCancel         |     Whenever the user cancels when changing the date.                                                |  void                                    |
|   list             |     List of Lists of Options to be displayed on pop up widget                                        |  required                                |
|   wtList           |     List flex propertry of visible widget items                                                      |  equal for each child                    |
|   contentPadding   |     The padding of the picker widget                                                                 |  zero                                    |
|   widDecoration    |     BoxDecoration for visible widget                                                                 |  BoxDecoration()                         |
|   childDecoration  |     BoxDecoration for visible widget's child                                                         |  default TextStyle                       |
|   widHeight        |     height for visible widget                                                                        |  40                                      |
|   divColor         |     Divider Color for visible Widget                                                                 |  Color(0xFF424242)                       |
|   popUpSelColor    |     PopUp Selected Item Color                                                                        |  Colors.white                            |


CHANGELOG.md
# CHANGES

## 0.0.1

- Released package!!

## 0.0.2

- Added more Customization in Text of Pop Up Menu
- Added more Decoration options


## 0.0.3
- Updated Documention

## 0.0.4
- added more customizations
- Updated Documention

test/custom_picker_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:custom_picker/custom_picker.dart';

void main() {
  test('adds one to input values', () {});
}


test/custom_dropdown_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:custom_picker/custom_dropdown.dart';

void main() {
  test('adds one to input values', () {});
}


lib/PopUpMenu.dart
import 'package:custom_picker/PopUpRoute.dart';
import 'package:custom_picker/PopUpScrollBehaviour.dart';
import 'package:flutter/material.dart';
import 'dart:math' as math;

class PopUpMenu<T> extends StatefulWidget {
  const PopUpMenu({
    Key? key,
    required this.route,
    required this.buttonRect,
    required this.constraints,
    this.dropdownColor,
  }) : super(key: key);

  final PopUpRoute<T> route;
  final Rect buttonRect;
  final BoxConstraints constraints;
  final Color? dropdownColor;

  @override
  _PopUpMenuState<T> createState() => _PopUpMenuState<T>();
}

class _PopUpMenuState<T> extends State<PopUpMenu<T>> {
  late CurvedAnimation _fadeOpacity;

  @override
  void initState() {
    super.initState();
    _fadeOpacity = CurvedAnimation(
      parent: widget.route.animation ?? const AlwaysStoppedAnimation<double>(1),
      curve: const Interval(0.0, 0.50),
      reverseCurve: const Interval(0.75, 1.0),
    );
  }

  @override
  Widget build(BuildContext context) {
    return FadeTransition(
      opacity: _fadeOpacity,
      child: Semantics(
        scopesRoute: true,
        namesRoute: true,
        explicitChildNodes: true,
        child: Material(
          type: MaterialType.transparency,
          child: ScrollConfiguration(
            behavior: const PopUpScrollBehavior(),
            child: widget.route.content,
          ),
        ),
      ),
    );
  }
}

class PopUpMenuRouteLayout<T> extends SingleChildLayoutDelegate {
  PopUpMenuRouteLayout({
    required this.buttonRect,
    required this.route,
    required this.textDirection,
    required this.target,
    required this.verticalOffset,
    this.width,
    this.below = false,
  });

  final Rect buttonRect;
  final PopUpRoute<T> route;
  final TextDirection textDirection;
  final Offset target;
  final bool below;
  final double verticalOffset;
  final double? width;

  @override
  BoxConstraints getConstraintsForChild(BoxConstraints constraints) {
    final double maxHeight = constraints.maxHeight;
    final double width =
        this.width ?? math.min(constraints.maxWidth, buttonRect.width);
    return BoxConstraints(
      minWidth: width,
      maxWidth: width,
      minHeight: 0.0,
      maxHeight: maxHeight,
    );
  }

  @override
  Offset getPositionForChild(Size size, Size childSize) {
    return positionDependentBox(
      size: size,
      childSize: childSize,
      target: target,
      preferBelow: below,
      verticalOffset: verticalOffset,
    );
  }

  @override
  bool shouldRelayout(PopUpMenuRouteLayout<T> oldDelegate) {
    return target != oldDelegate.target || buttonRect != oldDelegate.buttonRect;
  }
}


lib/PopUpRoute.dart
import 'package:custom_picker/PopUpMenu.dart';
import 'package:flutter/material.dart';

class PopUp<T> extends StatefulWidget {
  const PopUp({
    Key? key,
    required this.child,
    required this.content,
    this.contentHeight = 0,
    this.contentWidth,
    this.below,
    this.verticalOffset = 0,
  }) : super(key: key);

  final Widget child;
  final WidgetBuilder content;
  final double contentHeight;
  final double? contentWidth;
  final double verticalOffset;
  final bool? below;

  @override
  PopUpState<T> createState() => PopUpState<T>();
}

class PopUpState<T> extends State<PopUp<T>> {
  PopUpRoute<T>? _dropdownRoute;
  Future<void> openPopup() {
    final NavigatorState navigator = Navigator.of(context);
    final RenderBox itemBox = context.findRenderObject()! as RenderBox;
    final Offset target = itemBox.localToGlobal(
      itemBox.size.center(Offset.zero),
      ancestor: navigator.context.findRenderObject(),
    );
    final Rect itemRect = target & itemBox.size;
    assert(_dropdownRoute == null, 'You can NOT open it twice');
    _dropdownRoute = PopUpRoute<T>(
      width: widget.contentWidth,
      target: target,
      below: widget.below!,
      contentHeight: widget.contentHeight,
      content: widget.content(context),
      buttonRect: itemRect,
      elevation: 4,
      capturedThemes:
          InheritedTheme.capture(from: context, to: navigator.context),
      transitionAnimationDuration: Duration(milliseconds: 500),
      verticalOffset: widget.verticalOffset,
      barrierLabel: MaterialLocalizations.of(context).modalBarrierDismissLabel,
    );

    return navigator.push(_dropdownRoute!).then((T? newValue) {
      removePopUpRoute();
      if (!mounted || newValue == null) return;
    });
  }

  bool get isOpen => _dropdownRoute != null;

  void removePopUpRoute() {
    _dropdownRoute?._dismiss();
    _dropdownRoute = null;
  }

  @override
  void dispose() {
    removePopUpRoute();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return widget.child;
  }
}

class PopUpRoute<T> extends PopupRoute<T> {
  PopUpRoute({
    required this.content,
    required this.contentHeight,
    required this.buttonRect,
    required this.target,
    this.elevation = 8,
    required this.capturedThemes,
    required this.transitionAnimationDuration,
    required this.barrierLabel,
    this.width,
    this.below,
    required this.verticalOffset,
  });

  final Widget content;
  final double contentHeight;
  final Rect buttonRect;
  final int elevation;
  final CapturedThemes capturedThemes;
  final Offset target;
  final double? width;
  final bool? below;
  final double verticalOffset;

  final Duration transitionAnimationDuration;

  @override
  Duration get transitionDuration => transitionAnimationDuration;

  @override
  bool get barrierDismissible => true;

  @override
  Color? get barrierColor => null;

  @override
  final String barrierLabel;

  @override
  Widget buildPage(context, animation, secondaryAnimation) {
    return LayoutBuilder(builder: (context, constraints) {
      return _PopUpRoutePage<T>(
        target: target,
        route: this,
        below: below!,
        constraints: constraints,
        content: content,
        buttonRect: buttonRect,
        elevation: elevation,
        width: width,
        capturedThemes: capturedThemes,
        verticalOffset: verticalOffset,
      );
    });
  }

  void _dismiss() {
    if (isActive) {
      navigator?.removeRoute(this);
    }
  }
}

class _PopUpRoutePage<T> extends StatelessWidget {
  const _PopUpRoutePage({
    Key? key,
    required this.route,
    required this.constraints,
    required this.content,
    required this.buttonRect,
    required this.target,
    this.elevation = 8,
    required this.capturedThemes,
    required this.verticalOffset,
    this.style,
    this.below,
    this.width,
  }) : super(key: key);

  final PopUpRoute<T> route;
  final BoxConstraints constraints;
  final Widget content;
  final bool? below;
  final Rect buttonRect;
  final Offset target;
  final int elevation;
  final CapturedThemes capturedThemes;
  final TextStyle? style;
  final double? width;
  final double verticalOffset;

  @override
  Widget build(BuildContext context) {
    assert(debugCheckHasDirectionality(context));

    final TextDirection textDirection = TextDirection.ltr;
    final Widget menu = PopUpMenu<T>(
      route: route,
      buttonRect: buttonRect,
      constraints: constraints,
    );

    return MediaQuery.removePadding(
      context: context,
      removeTop: true,
      removeBottom: true,
      removeLeft: true,
      removeRight: true,
      child: Builder(
        builder: (BuildContext context) {
          return CustomSingleChildLayout(
            delegate: PopUpMenuRouteLayout<T>(
              target: target,
              below: below!,
              buttonRect: buttonRect,
              route: route,
              textDirection: textDirection,
              verticalOffset: verticalOffset,
              width: width,
            ),
            child: capturedThemes.wrap(menu),
          );
        },
      ),
    );
  }
}


lib/YesNoDialog.dart
import 'package:flutter/material.dart';

class YesNoPickerControl extends StatelessWidget {
  const YesNoPickerControl({
    Key? key,
    required this.onChanged,
    required this.onCancel,
    this.color = const Color(0xFFB9B6B6),
  }) : super(key: key);

  final VoidCallback onChanged;
  final VoidCallback onCancel;
  final Color? color;

  @override
  Widget build(BuildContext context) {
    return Row(
      children: [
        Expanded(
          child: SizedBox(
            height: 50,
            child: MaterialButton(
              child: Icon(
                Icons.check,
                color: color,
              ),
              onPressed: onChanged,
            ),
          ),
        ),
        Expanded(
          child: SizedBox(
            height: 50,
            child: MaterialButton(
              child: Icon(
                Icons.close,
                color: color,
              ),
              onPressed: onCancel,
            ),
          ),
        ),
      ],
    );
  }
}


lib/custom_picker.dart
library custom_picker;

import 'package:custom_picker/CustomPickerContentPopUp.dart';
import 'package:custom_picker/PopUpRoute.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

class CustomPicker extends StatefulWidget {
  const CustomPicker({
    Key? key,
    required this.selected,
    required this.onChanged,
    this.handleChange,
    this.onCancel,
    this.verticalOffset = 0,
    this.below = true,
    required this.list,
    this.optionLoop,
    this.contentPadding = const EdgeInsets.all(0),
    this.popupHeight = 400,
    this.divColor = const Color(0xFFB9B6B6),
    this.wtList,
    this.widDecoration = const BoxDecoration(),
    this.childDecoration = const [],
    this.widHeight = 40,
    this.yesNo = false,
    this.popUpMargin,
    this.popUpSelColor,
    this.popUpDecoration,
    this.popUpTextStyle,
    this.widTextStyle,
  }) : assert((list.length == selected.length &&
                selected.length == childDecoration.length) ||
            (list.length == selected.length &&
                (childDecoration == null || childDecoration.length == 0)));

  /// give a decoration to your popupDecpration
  final Decoration? popUpDecoration;

  /// vertical position of popup widget
  final double verticalOffset;

  /// show popup widget below the [CustomPicker]
  final bool below;

  /// initially selected list
  final List<int> selected;

  /// whether to show yes no dialog
  final bool yesNo;

  /// add TextStyle of popup children
  final List<TextStyle>? popUpTextStyle;

  /// add TextStyles of Picker Children
  final List<TextStyle>? widTextStyle;

  /// Whenever the current date is changed. If this is null, the picker is considered disabled
  final ValueChanged<List<int>> onChanged;

  /// any change in any child will call this method
  final ValueChanged<List<int>>? handleChange;

  /// Margin for PopUp Widget (if it is zero, pop up widget will be visible at the center of picker)
  final EdgeInsets? popUpMargin;

  /// Whenever the user cancels when changing the date.
  final VoidCallback? onCancel;

  ///List<List<dynamic>> List of Lists of Options to be displayed on pop up widget
  final List<List<dynamic>> list;

  ///List<int> flex propertry of visible widget items
  final List<int>? wtList;

  /// The padding of the picker
  final EdgeInsetsGeometry contentPadding;

  ///BoxDecoration for visible widget
  final BoxDecoration widDecoration;

  ///BoxDecoration for visible widget's child
  final List<BoxDecoration> childDecoration;

  ///height for visible widget
  final double widHeight;

  /// The height of the popup
  final double popupHeight;

  /// Divider Color for visible Widget
  final Color divColor;

  /// PopUp Selected Item Color
  final Color? popUpSelColor;

  ///List<bool> which child to loop over wheel
  final List<bool>? optionLoop;

  @override
  _CustomPickerState createState() => _CustomPickerState();
}

class _CustomPickerState extends State<CustomPicker> {
  List<int> sel = [];
  final popupKey = GlobalKey<PopUpState>();
  List<FixedExtentScrollController>? controllers = [];

  @override
  void initState() {
    super.initState();
    widget.selected.forEach((e) => sel.add(e));
    initControllers();
  }

  void initControllers() {
    for (int i = 0; i < widget.list.length; i++) {
      controllers?.add(
        FixedExtentScrollController(
          initialItem: sel[i],
        ),
      );
    }
  }

  @override
  void dispose() {
    for (int i = 0; i < widget.list.length; i++) {
      controllers?[i].dispose();
      controllers = null;
    }
    super.dispose();
  }

  @override
  void didUpdateWidget(CustomPicker oldWidget) {
    super.didUpdateWidget(oldWidget);
    bool isTrue = false;
    if (sel.length == widget.selected.length) {
      for (int i = 0; i < widget.selected.length; i++) {
        if (widget.selected[i] != sel[i]) {
          isTrue = true;
        }
      }
    } else {
      isTrue = true;
    }
    if (isTrue) {
      sel.clear();
      widget.selected.forEach((e) => sel.add(e));
      for (int i = 0; i < widget.list.length; i++) {
        controllers?[i].jumpToItem(sel[i]);
      }
    }
  }

  Future<void> handleDataChanged(List<int> list) async {
    widget.handleChange?.call(list);
  }

  @override
  Widget build(BuildContext context) {
    bool style = (widget.widTextStyle != null) &&
        (widget.widTextStyle?.length == widget.selected.length);
    Widget picker = MaterialButton(
        onPressed: () async {
          await popupKey.currentState?.openPopup();
          for (int i = 0; i < widget.list.length; i++) {
            controllers?[i].dispose();
          }
          controllers = [];
          initControllers();
        },
        child: AnimatedContainer(
            duration: const Duration(milliseconds: 700),
            decoration: widget.widDecoration,
            curve: Curves.bounceIn,
            height: widget.widHeight,
            child: Row(
              children: List.generate(
                widget.list.length,
                (index) {
                  TextStyle styled =
                      style ? widget.widTextStyle![index] : TextStyle();
                  return Expanded(
                    flex: widget.wtList == null ? 1 : widget.wtList![index],
                    child: Padding(
                      padding: widget.contentPadding,
                      child: Row(
                        children: [
                          Expanded(
                            child: Container(
                              decoration: widget.childDecoration != null &&
                                      widget.childDecoration.length > 0
                                  ? widget.childDecoration[index]
                                  : BoxDecoration(),
                              child: Center(
                                child: Text(
                                  '${widget.list[index][widget.selected[index]]}',
                                  style: styled,
                                  textAlign: TextAlign.center,
                                ),
                              ),
                            ),
                          ),
                          SizedBox(
                              width: index != widget.list.length - 1 ? 15 : 0),
                          index != widget.list.length - 1
                              ? VerticalDivider(
                                  color: widget.divColor,
                                  thickness: 1.5,
                                  width: 0,
                                )
                              : SizedBox(),
                        ],
                      ),
                    ),
                  );
                },
              ),
            )));
    picker = PopUp(
      contentHeight: widget.popupHeight,
      key: popupKey,
      child: picker,
      below: widget.below,
      verticalOffset: widget.verticalOffset,
      content: (context) => CustomPickerContentPopUp(
        popUpMargin: widget.popUpMargin,
        popUpSelColor: widget.popUpSelColor,
        height: widget.popupHeight,
        popUpTextStyle: widget.popUpTextStyle,
        popUpDecoration: widget.popUpDecoration,
        selected: sel,
        data: widget.list,
        optionLoop: widget.optionLoop,
        controllers: controllers!,
        yesNo: widget.yesNo,
        handleChanged: handleDataChanged,
        onCancel: () => widget.onCancel?.call(),
        onChanged: () => widget.onChanged.call(sel),
      ),
    );
    return picker;
  }
}


lib/CustomPickerContentPopUp.dart
import 'package:custom_picker/YesNoDialog.dart';
import 'package:flutter/material.dart';

class CustomPickerContentPopUp extends StatefulWidget {
  const CustomPickerContentPopUp({
    Key? key,
    required this.handleChanged,
    required this.onChanged,
    required this.onCancel,
    required this.height,
    required this.selected,
    required this.controllers,
    required this.data,
    this.optionLoop,
    this.yesNo,
    this.popUpMargin,
    this.popUpSelColor,
    this.popUpDecoration,
    this.popUpTextStyle,
  }) : super(key: key);

  final bool? yesNo;
  final List<int> selected;
  final ValueChanged<List<int>> handleChanged;
  final VoidCallback onChanged;
  final VoidCallback onCancel;
  final double height;
  final EdgeInsets? popUpMargin;
  final List<bool>? optionLoop;
  final List<FixedExtentScrollController> controllers;
  final List<List<dynamic>> data;
  final Color? popUpSelColor;
  final Decoration? popUpDecoration;
  final List<TextStyle>? popUpTextStyle;

  @override
  CustomPickerContentPopUpState createState() =>
      CustomPickerContentPopUpState();
}

class CustomPickerContentPopUpState extends State<CustomPickerContentPopUp> {
  @override
  Widget build(BuildContext context) {
    bool loop = (widget.optionLoop != null &&
        widget.optionLoop?.length == widget.selected.length);
    bool style = (widget.popUpTextStyle != null &&
        widget.popUpTextStyle?.length == widget.selected.length);
    final divider = VerticalDivider(
      color: Color(0xFFB9B6B6),
      thickness: 2,
    );
    return Container(
      margin: widget.popUpMargin != null
          ? widget.popUpMargin
          : const EdgeInsets.symmetric(horizontal: 20, vertical: 30),
      height: widget.height,
      decoration: widget.popUpDecoration ??
          BoxDecoration(
            color: const Color(0xFFC4C4C4),
            borderRadius: BorderRadius.circular(10),
            border: Border.all(
              width: 2,
              color: const Color(0xFFB9B6B6),
            ),
          ),
      child: Column(
        children: [
          Expanded(
            child: Stack(
              children: [
                Center(
                  child: Container(
                    alignment: Alignment.center,
                    height: 40,
                    color: widget.popUpSelColor ?? Colors.white,
                  ),
                ),
                Row(
                  children: List.generate(
                    widget.data.length,
                    (index) {
                      bool looping = loop ? widget.optionLoop![index] : true;
                      TextStyle styled = style
                          ? widget.popUpTextStyle![index]
                          : TextStyle(fontSize: 17);
                      return Expanded(
                        flex: 2,
                        child: ListWheelScrollView.useDelegate(
                          controller: widget.controllers[index],
                          itemExtent: 40,
                          diameterRatio: 10,
                          renderChildrenOutsideViewport: false,
                          physics: const FixedExtentScrollPhysics(),
                          childDelegate: looping
                              ? ListWheelChildLoopingListDelegate(
                                  children: List.generate(
                                    widget.data[index].length,
                                    (ind) {
                                      final text = "${widget.data[index][ind]}";
                                      return Center(
                                        child: Text(
                                          text,
                                          style: styled,
                                        ),
                                      );
                                    },
                                  ),
                                )
                              : ListWheelChildListDelegate(
                                  children: List.generate(
                                    widget.data[index].length,
                                    (ind) {
                                      final text = "${widget.data[index][ind]}";
                                      return Center(
                                        child: Text(
                                          text,
                                          style: styled,
                                        ),
                                      );
                                    },
                                  ),
                                ),
                          onSelectedItemChanged: (change) {
                            widget.selected[index] = change;
                            widget.handleChanged(widget.selected);
                            setState(() {});
                          },
                        ),
                      );
                    },
                  ),
                ),
              ],
            ),
          ),
          widget.yesNo!
              ? Divider(
                  color: widget.popUpSelColor ?? Color(0xFFB9B6B6),
                  thickness: 1.5,
                  height: 0,
                )
              : SizedBox(),
          widget.yesNo!
              ? YesNoPickerControl(
                  color: widget.popUpSelColor,
                  onChanged: () {
                    Navigator.pop(context);
                    widget.onChanged.call();
                  },
                  onCancel: () {
                    Navigator.pop(context);
                    widget.onCancel();
                  },
                )
              : SizedBox(),
        ],
      ),
    );
  }
}


example/lib/main.dart
import 'package:custom_picker/custom_picker.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  var opt = [5, 17, 41];
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Column(
          mainAxisAlignment: MainAxisAlignment.end,
          children: [
            CustomPicker(
              // deciding list whether to show loop over popup items
              optionLoop: [false, false, false],
              // give popup height
              popupHeight: 500,
              // where to show popupmenu
              below: false,
              // give where to place popup vertically
              verticalOffset: 10,
              // hanfle all changes here
              handleChange: (list) {
                print(list);
              },
              // to include yes no dialog or not
              yesNo: true,
              //give color to popup selected option
              popUpSelColor: Colors.black.withAlpha(50),
              // customize text style for each popUp list
              popUpTextStyle: [
                TextStyle(color: Colors.grey.shade600, fontSize: 18),
                TextStyle(color: Colors.grey.shade600, fontSize: 18),
                TextStyle(color: Colors.grey.shade600, fontSize: 18),
              ],
              // customize text style for each list
              widTextStyle: [
                TextStyle(color: Colors.black.withAlpha(120), fontSize: 18),
                TextStyle(color: Colors.black.withAlpha(120), fontSize: 18),
                TextStyle(color: Colors.black.withAlpha(120), fontSize: 18),
              ],
              // customize divider color
              divColor: Colors.grey.shade200,
              //add padding
              contentPadding: const EdgeInsets.all(7),
              //give a decoration to your popupmenu
              popUpDecoration: BoxDecoration(
                  color: Colors.white,
                  border: Border.all(width: 2, color: Colors.white),
                  boxShadow: [
                    BoxShadow(
                      color: const Color(0x33000000).withAlpha(50),
                      offset: Offset(2, 4),
                      blurRadius: 7,
                    ),
                  ],
                  borderRadius: BorderRadius.circular(10)),
              // decorate your picker widget
              widDecoration: BoxDecoration(
                color: Colors.white,
                boxShadow: [
                  BoxShadow(
                    color: const Color(0x33000000).withAlpha(50),
                    offset: Offset(2, 4),
                    blurRadius: 7,
                  ),
                ],
                borderRadius: BorderRadius.circular(20),
              ),
              // input all list you want to display in popupmenu
              list: [
                [
                  'Jan',
                  'Feb',
                  'Mar',
                  'Apr',
                  'May',
                  'Jun',
                  'Jul',
                  'Aug',
                  'Sep',
                  'Oct',
                  'Nov',
                  'Dec'
                ],
                List.generate(31, (index) => '${index + 1}').toList(),
                List.generate(50, (index) => '${index + 1980}').toList(),
              ],
              // control flex property of picker children
              wtList: [3, 1, 2],
              // callback that'll be call when option of YesNo dialog is clicked
              onChanged: (List<int> value) {
                opt = value;
                setState(() {});
              },
              //give initial selected item indicies
              selected: opt,
            ),
            SizedBox(height: 100),
          ],
          // CustomPicker(
          //   // deciding list whether to show loop over popup items
          //   optionLoop: [false, false],
          //   // give popup height
          //   popupHeight: 200,
          //   // where to show popupmenu
          //   below: false,
          //   // give where to place popup vertically
          //   verticalOffset: 10,
          //   // hanfle all changes here
          //   handleChange: (list) {
          //     print(list);
          //   },
          //   // to include yes no dialog or not
          //   yesNo: true,
          //   //give color to popup selected option
          //   popUpSelColor: Colors.black.withAlpha(150),
          //   // customize text style for each popUp list
          //   popUpTextStyle: const [
          //     TextStyle(color: Colors.white, fontSize: 20),
          //     TextStyle(color: Colors.white, fontSize: 20),
          //   ],
          //   // customize text style for each list
          //   widTextStyle: const [
          //     TextStyle(color: Colors.black, fontSize: 18),
          //     TextStyle(color: Colors.black, fontSize: 18),
          //   ],
          //   // customize divider color
          //   divColor: Colors.black,
          //   //give a decoration to your popupmenu
          //   popUpDecoration: BoxDecoration(
          //       color: Colors.black.withAlpha(100),
          //       borderRadius: BorderRadius.circular(10)),
          //   // decorate your picker widget
          //   widDecoration: BoxDecoration(
          //     border: Border.all(width: 2, color: Colors.black),
          //     borderRadius: BorderRadius.circular(10),
          //   ),
          //   // input all list you want to display in popupmenu
          //   list: const [
          //     ['1', '2', '3', '4', '5', '6'],
          //     ['one', 'two', 'three']
          //   ],
          //   // control flex property of picker children
          //   wtList: [3, 1],
          //   // callback that'll be call when option of YesNo dialog is clicked
          //   onChanged: (List<int> value) {
          //     print(value);
          //   },
          //   //give initial selected item indicies
          //   selected: const [0, 1],
          // ),
        ),
      ),
    );
  }
}
